# 第1章 性能监视工具

## 1.1 计算素数

**程序P1：** 依次打印所有小于1000的素数。

In [1]:
from line_profiler import LineProfiler

def prime(n):
    for i in range(2, n):
        if n % i == 0:
            return 0
    return 1

def main():
    res = []
    n = 1000
    for i in range(2, n + 1):
        if prime(i):
            res.append(i)     
    return res

In [2]:
p = LineProfiler()
p.add_function(prime)
p_warp = p(main)
p_warp()
p.print_stats()

Timer unit: 1e-07 s

Total time: 0.0388577 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\1730646751.py
Function: prime at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           def prime(n):
     4     78190     162787.0      2.1     41.9      for i in range(2, n):
     5     78022     223515.0      2.9     57.5          if n % i == 0:
     6       831       1872.0      2.3      0.5              return 0
     7       168        403.0      2.4      0.1      return 1

Total time: 0.0688222 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\1730646751.py
Function: main at line 9

Line #      Hits         Time  Per Hit   % Time  Line Contents
     9                                           def main():
    10         1          5.0      5.0      0.0      res = []
    11         1          4.0      4.0      0.0      n = 1000
    12      1000       2410.0      2.4      0.4      for i in range(2, n + 1):
    1

测试结论：上述main函数调用了1次，测试了999个整数，找出了168个素数，prime调用了999次，其中168次返回1，831次返回0，测试了78022个因子，每个整数检查了大约78个因子。

**程序P2：** 对$n$考虑不超过$\sqrt{n}$的可能的整数因子。

In [3]:
import math

from line_profiler import LineProfiler


def root(n):
    return int(math.sqrt(n))


def prime(n):
    for i in range(2, root(n) + 1):
        if n % i == 0:
            return 0
    return 1


def main():
    res = []
    n = 1000
    for i in range(2, n + 1):
        if prime(i):
            res.append(i)
    return res

In [4]:
p = LineProfiler()
p.add_function(prime)
p_warp = p(main)
p_warp()
p.print_stats()

Timer unit: 1e-07 s

Total time: 0.0032815 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\3566730837.py
Function: prime at line 10

Line #      Hits         Time  Per Hit   % Time  Line Contents
    10                                           def prime(n):
    11      5456      18016.0      3.3     54.9      for i in range(2, root(n) + 1):
    12      5288      12907.0      2.4     39.3          if n % i == 0:
    13       831       1580.0      1.9      4.8              return 0
    14       168        312.0      1.9      1.0      return 1

Total time: 0.0062173 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\3566730837.py
Function: main at line 17

Line #      Hits         Time  Per Hit   % Time  Line Contents
    17                                           def main():
    18         1          5.0      5.0      0.0      res = []
    19         1          2.0      2.0      0.0      n = 1000
    20      1000       2000.0      2.0      3.2      for i in range(2, n

**程序P3：** 将sqrt调用移到循环之外

In [5]:
def prime(n):
    bound = root(n)
    for i in range(2, bound + 1):
        if n % i == 0:
            return 0
    return 1

In [6]:
p = LineProfiler()
p.add_function(prime)
p_warp = p(main)
p_warp()
p.print_stats()

Timer unit: 1e-07 s

Total time: 0.0034453 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\2702493981.py
Function: prime at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def prime(n):
     2       999       8079.0      8.1     23.4      bound = root(n)
     3      5456      12028.0      2.2     34.9      for i in range(2, bound + 1):
     4      5288      12337.0      2.3     35.8          if n % i == 0:
     5       831       1696.0      2.0      4.9              return 0
     6       168        313.0      1.9      0.9      return 1

Total time: 0.0067002 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\3566730837.py
Function: main at line 17

Line #      Hits         Time  Per Hit   % Time  Line Contents
    17                                           def main():
    18         1          5.0      5.0      0.0      res = []
    19         1          3.0      3.0      0.0      n = 1000
    20

**程序P4：** 对2、3和5整除的特殊检验，但是结果不正确。

In [7]:
def prime(n):
    if n % 2 == 0:
        return 0
    if n % 3 == 0:
        return 0
    if n % 5 == 0:
        return 0
    bound = root(n)
    for i in range(7, bound + 1, 2):
        if n % i == 0:
            return 0
    return 1

In [8]:
p = LineProfiler()
p.add_function(prime)
p_warp = p(main)
p_warp()
p.print_stats()

Timer unit: 1e-07 s

Total time: 0.0017084 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\349108320.py
Function: prime at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def prime(n):
     2       999       2635.0      2.6     15.4      if n % 2 == 0:
     3       500        986.0      2.0      5.8          return 0
     4       499       1294.0      2.6      7.6      if n % 3 == 0:
     5       167        322.0      1.9      1.9          return 0
     6       332        794.0      2.4      4.6      if n % 5 == 0:
     7        67        127.0      1.9      0.7          return 0
     8       265       2379.0      9.0     13.9      bound = root(n)
     9      1695       4068.0      2.4     23.8      for i in range(7, bound + 1, 2):
    10      1530       3904.0      2.6     22.9          if n % i == 0:
    11       100        229.0      2.3      1.3              return 0
    12       165        346.0     

**程序P5:** 把开方运算换成乘法，对被2、3、5整除的正确检验。

In [11]:
def prime(n):
    if n % 2 == 0:
        return n == 2
    if n % 3 == 0:
        return n == 3
    if n % 5 == 0:
        return n == 5
    i = 7
    while i * i <= n:
        if n % i == 0:
            return 0
        i += 2

    return 1

In [12]:
p = LineProfiler()
p.add_function(prime)
p_warp = p(main)
p_warp()
p.print_stats()

Timer unit: 1e-07 s

Total time: 0.004503 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\3566730837.py
Function: main at line 17

Line #      Hits         Time  Per Hit   % Time  Line Contents
    17                                           def main():
    18         1          6.0      6.0      0.0      res = []
    19         1          3.0      3.0      0.0      n = 1000
    20      1000       2190.0      2.2      4.9      for i in range(2, n + 1):
    21       999      42226.0     42.3     93.8          if prime(i):
    22       168        603.0      3.6      1.3              res.append(i)
    23         1          2.0      2.0      0.0      return res

Total time: 0.0019244 s
File: C:\Users\hurui\AppData\Local\Temp\ipykernel_26816\4250436128.py
Function: prime at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def prime(n):
     2       999       2653.0      2.7     13.8      if n % 2 == 0:
     3 

## 1.2 使用性能监视工具

> Don Kunth的一篇论文中写道：“一个程序中不到4%的语句通常占用了一半以上的运行时间”

对于不处理I/O密集型的大多数程序，大部分的运行时间花在了很小的一部分代码上。

## 1.3 原理

- 使用性能监视工具
- 开发性能监视工具：在Python中，可以使用IDE自带的Profile，适用于调用链的排查；可以应用line_profile包，适用于模块化代码的性能分析。